Enhanced Location Search with Department Support

ledjay opened the pull request:
# 🔍 Enhanced Location Search with Department Support

## What's New

This PR supercharges the location search on the `/recherche` page by enabling **dual-API search** that queries both cities and departments simultaneously, giving users more flexibility and better results.

### Key Changes

- **🏙️ City + Department Search**: Auto-suggest now searches both municipalities and departments in parallel, merging results intelligently
- **🏷️ Better Labels**: Changed default filter label from "Département" to "Ville" to reflect the primary search use case
- **🚀 Smart Relevance**: Results sorted by relevance with normalized string matching (handles accents, hyphens, case)
- **⚠️ Deployment Awareness**: Warns users when selecting departments that aren't yet deployed
- **🎯 Type Safety**: Added proper TypeScript types for unified search results

### Technical Highlights

- Dual API integration: `geo.api.gouv.fr` (municipalities) + departments API
- New helper functions in `LocationMenu/functions.ts` for result transformation and sorting
- Defensive programming: Added `?? []` fallback in [activeDispositifsSelector](cci:1://file:///Users/jeremie/refugies.info/karfur/apps/client/src/services/ActiveDispositifs/activeDispositifs.selector.ts:3:0-3:113) to prevent null crashes
- Comprehensive spec documentation in `/specs/001-enhance-location-search/`

## 🧪 How to Test

### Test 1: City Search (Primary Use Case)
1. Navigate to `/recherche`
2. Click on the location filter (should show "Ville" as default label)
3. Type **"Paris"** in the search field
4. ✅ Verify you see both:
   - City result: "Paris (75)"
   - Department result: "Paris (dept)"
5. Select "Paris (75)" (city)
6. ✅ Verify the filter button updates to show "Paris"
7. ✅ Check URL contains `?departments={department}`

### Test 2: Department Search
1. Clear the location filter
2. Type **"Rhône"** in the search field
3. ✅ Verify you see "Rhône (dept)" in results
4. Select it
5. ✅ Verify filter button shows "Rhône"
6. ✅ Check URL contains `?departments={department}`

### Test 3: Undeployed Department Warning
1. Type **"Corse"** or another undeployed department
2. Select it
3. ✅ Verify you see a warning message about limited content availability

### Test 4: Accent & Hyphen Handling
1. Type **"rhone"** (without accent)
2. ✅ Verify "Rhône" still appears in results
3. Type **"saint etienne"** (without hyphen)
4. ✅ Verify "Saint-Étienne" appears in results

### Test 5: Mobile Experience
1. Open `/recherche` on mobile viewport (< 768px)
2. Open location filter
3. ✅ Verify search works smoothly
4. ✅ Verify labels are appropriate for small screens

### Test 6: Accessibility
1. Use keyboard only (Tab, Enter, Escape)
2. ✅ Verify you can navigate and select locations
3. Use screen reader
4. ✅ Verify announcements are clear when results update

### Test 7: Regression Check
1. Test existing location filter functionality
2. ✅ Verify "common places" still show when search is empty
3. ✅ Verify Redux state updates correctly
4. ✅ Verify no console errors

## 📋 Files Changed

**Core Implementation:**
- [LocationMenu/LocationMenu.tsx](cci:7://file:///Users/jeremie/refugies.info/karfur/apps/client/src/components/Pages/recherche/LocationMenu/LocationMenu.tsx:0:0-0:0) - Dual API search logic
- `LocationMenu/functions.ts` - Helper functions (NEW)
- `LocationMenu/SearchMenuItem.tsx` - Result rendering
- [SearchHeader/Filters.tsx](cci:7://file:///Users/jeremie/refugies.info/karfur/apps/client/src/components/Pages/recherche/SearchHeader/Filters.tsx:0:0-0:0) - Label update

**Utilities:**
- `lib/recherche/functions.ts` - Department deployment check
- [services/ActiveDispositifs/activeDispositifs.selector.ts](cci:7://file:///Users/jeremie/refugies.info/karfur/apps/client/src/services/ActiveDispositifs/activeDispositifs.selector.ts:0:0-0:0) - Null safety

**Translations:**
- Updated all locale files with new "Ville" label

**Documentation:**
- Complete spec suite in `/specs/001-enhance-location-search/`

Fixes RI-907

ledjay committed:
feat: improve location search with department support

- Added department search alongside city search to support both search types
- Updated placeholder text and labels to reflect combined city/department search
- Changed "Department" filter label to "City" in UI translations
- Removed redundant department filter labels in non-French translations
- Added department deployment status check and user notifications
- Improved search relevance with normalized string matching for accents/hyphens
- Added type

kaaloo reviewed (changes requested):
All tests passed but I noticed the following:

In mobile view, view and theme button are spread out:

<img width="735" height="510" alt="Image" src="https://github.com/user-attachments/assets/0cca5d6d-dc14-47dc-b53f-0f5ba0c0b10b" />

I also see the following in the console, maybe not related:

```
hook.js:608 Warning: Missing `Description` or `aria-describedby={undefined}` for {DialogContent}. Error Component Stack
    at DialogMenuLayout (DialogMenuLayout.tsx:15:36)
    at div (<anonymous>)
    at Filter (Filter.tsx:71:3)
    at div (<anonymous>)
    at DropdownProvider (DropDownMenuContext.tsx:14:69)
    at div (<anonymous>)
    at Filters (Filters.tsx:33:11)
    at div (<anonymous>)
    at header (<anonymous>)
    at SearchHeader (SearchHeader.tsx:25:31)
    at div (<anonymous>)
    at Recherche (recherche.tsx:76:31)
    at main (<anonymous>)
    at div (<anonymous>)
    at div (<anonymous>)
    at Layout (Layout.tsx:51:57)
    at ScreenReaderAnnouncerProvider (ScreenReaderAnnouncer.tsx:28:49)
    at App (_app.tsx:53:16)

```

ledjay committed:
feat: improve location filtering and UI in search menu

- Updated location filter UI from radio buttons to checkboxes to allow multiple department selection
- Added clear filter announcement when removing last department selection
- Fixed French translations for "sélectionné" and added "departmentsCleared" message
- Enhanced location menu styling with better spacing and scrollable results
- Improved department name handling with proper decoding of special characters
- Added visual indication of

ledjay committed:
feat: improve accessibility and mobile layout for search filters

- Added aria-describedby and description text for filter dialog to improve screen reader support
- Added aria-describedby for language selector dialog to enhance accessibility
- Modified location menu CSS to remove max-height limit on mobile for better usability
- Added French translation for filter dialog description
- Optimized location menu scrolling behavior to only apply on desktop (>= 768px)

ledjay committed:
refactor: simplify location search sorting logic

- Extracted normalizeString function to be reusable across components
- Refactored sortByRelevance function to use a score-based approach instead of nested conditions
- Simplified sorting logic by using a getScore helper function with clear priority levels
- Removed max-height and overflow CSS constraints from location menu
- Added clear scoring documentation for different match types (exact match, starts with, contains)

The changes improve code

ledjay committed:
feat: enhance location filtering with city-level selection

- Added city-level filtering alongside department filtering to allow more granular location selection
- Created new section to display currently selected locations (departments and cities)
- Added automatic department selection when a city is chosen
- Updated French translations to support city-specific labels and messages
- Added scroll behavior and improved styling for location menu container
- Improved accessibility by updating ARIA announcements for

ledjay committed:
feat: update location filter terminology in search UI

- Changed "localité" to "localisation" across search components for more accurate French terminology
- Updated translation keys and display text in LocationMenu component
- Modified filter labels and announcement messages in SearchHeader for consistency
- Standardized terminology in location-related UI elements to improve user comprehension

ledjay committed:
feat: update French translations for location terminology

- Changed "localité" to "localisation" in multiple filter labels and messages for consistency
- Updated translations for department filter clear message and selected locations text
- Standardized location-related terminology throughout the French locale file to use "localisation"

ledjay committed:
feat: make cities filter optional in search query

- Updated SearchQuery type to make cities array an optional parameter by adding '?' modifier
- Allows search functionality to work without requiring city filters to be specified
- Maintains backwards compatibility while adding more flexibility to search queries
